### PR TITLE
Add Aktualizr (SOTA Client written in C++)

### PR DESCRIPTION
**Pull request by leon-anavi**
Hi,

These patches replace RVI SOTA Client written in the Rust programming language with Aktualizr which is written in C++. Summary of the provided patches:
* Remove the old RVI SOTA Client written in Rust
* Add Aktualizr and a systemd service for it
* Add rvi-sample-config which provides configurations, including a sample sota.toml
* Add Aktualizr and rvi-sample-config to packagegroup-gdp-rvi to ensure that they will be include in image genivi-dev-platform

Best regards,
Leon

### CODE REVIEW

**Comment by zeenix** [comment]
@leon-anavi Really good stuff! You'll be happy to hear that there are only few minor comments I'll post on individual commits. :)

**Review by zeenix — CHANGES_REQUESTED** [review]
Re: "rvi-sample-config: Add recipe"

Commit log is pretty good but can you please also add a bit to the description so it says what sample configuration commit is about. Bonus points for also describing the need for this change in particular.

File: meta-genivi-dev/recipes-sota/aktualizr/files/aktualizr.service
Comment (zeenix) [inline meta-genivi-dev/recipes-sota/aktualizr/files/aktualizr.service]: So sorry that I have to go back on what I just said about everything being perfect but why is the service running on session bus? Session is for services that run for each user session separately and AFAIK SOTA is not meant to be per-user service. Besides we are installing it as a system service so that makes it very inconsistent.

**Comment by leon-anavi** [comment]
Hi @zeenix, thank you for the feedback. Following your recommendations I have updated my patches. 

Best regards,
Leon

File: meta-genivi-dev/recipes-sota/aktualizr/aktualizr_git.bb
Comment (chbae) [inline meta-genivi-dev/recipes-sota/aktualizr/aktualizr_git.bb]: You don't need this line

File: meta-genivi-dev/recipes-sota/aktualizr/aktualizr_git.bb
Comment (chbae) [inline meta-genivi-dev/recipes-sota/aktualizr/aktualizr_git.bb]: Why do you add dependency with library?
If you really required, please use "RDEPENDS_${PN} = "dbus-lib"

File: meta-genivi-dev/recipes-sota/aktualizr/aktualizr_git.bb
Comment (chbae) [inline meta-genivi-dev/recipes-sota/aktualizr/aktualizr_git.bb]: Please use 4 spaces

**Review by chbae — CHANGES_REQUESTED** [review]

File: meta-genivi-dev/recipes-sota/rvi-sample-config/rvi-sample-config.bb
Comment (chbae) [inline meta-genivi-dev/recipes-sota/rvi-sample-config/rvi-sample-config.bb]: PR = "r0"

File: meta-genivi-dev/recipes-sota/rvi-sample-config/rvi-sample-config.bb
Comment (chbae) [inline meta-genivi-dev/recipes-sota/rvi-sample-config/rvi-sample-config.bb]: please use 4 spaces

**Review by chbae — CHANGES_REQUESTED** [review]

File: meta-genivi-dev/recipes-sota/aktualizr/aktualizr_git.bb
Comment (chbae) [inline meta-genivi-dev/recipes-sota/aktualizr/aktualizr_git.bb]: DEPENDS = "boost curl openssl jansson dbus"

**Comment by chbae** [comment]
@leon-anavi 
Do we still require meta-rust layer after merging this PR?

**Comment by gunnarx** [comment]
Build checks: Dragonboard failure is unrelated - bug filed and fix is coming. You can ignore it for this PR.

File: meta-genivi-dev/recipes-sota/aktualizr/aktualizr_git.bb
Comment (zeenix) [reply meta-genivi-dev/recipes-sota/aktualizr/aktualizr_git.bb]: hmm.. talking of which, is it really a runtime dep? Unless it's loaded dynamically (like a plugin), I doubt it's the case.

**Comment by leon-anavi** [comment]
Hi @zeenix, @chbae,

Thank you for the feedback. I have updated the GitHub pull request. 

@chbae, layer meta-rust will NOT be needed any more after merging this GitHub pull request.

Best regards,
Leon

**Review by zeenix — COMMENTED** [review]
Would be nice to have this working out of the box, so SOTA can finally be demoed out of the box in GDP. So it would be very nice if instead of adding sample configuration, we add default configuration.

@cajun-rat ^

**Comment by cajun-rat** [comment]
@zeenix The difficulty here is that we need some way to link a device to an individual user account: when a device comes on-line for the first time it needs to know whether it belongs to my sota.genivi.org account or yours.

Our current plan is that the process will be a simple matter of renaming sota.toml.example to sota.toml and changing the device ID to match the ID from the UI. I think the best thing here is to create some user documentation to explain this process--does that make sense?

**Comment by chbae** [comment]
@leon-anavi That's really good. rust is really heavy to build. It looks fine. Please provide how to test it.

**Comment by zeenix** [comment]
@cajun-rat Thanks. Yes, if we can make the process as easy as possible and document it in very simple instructions, we're mostly there already.

**Comment by zeenix** [comment]
@chbae yeah but if we needed/wanted to keep rust, we should have looked at removing it from the build and depend on rustc from host.

**Comment by leon-anavi** [comment]
Hi @zeenix,

It seems that you are happy with @cajun-rat proposal for writing user documentation to explain the usage of Aktualizr. If there are no technical remarks about the patches could you please merge them? :)

After merging these patches we can proceed with removing meta-rust and updating meta-genivi-dev in GDP git repo as well as with writing the user documentation in GENIVI wiki.

Thanks,
Leon

**Comment by zeenix** [comment]
@leon-anavi 

> It seems that you are happy with @cajun-rat proposal for writing user documentation to explain the usage of Aktualizr. If there are no technical remarks about the patches could you please merge them? :)

I think the understanding is that the sample configuration becomes the default configuration, to keep the instructions to setup a demo very simple and straightforward. Unless I misunderstood?

**Review by gunnarx — APPROVED** [review]
Looks good to me

**Review by gunnarx — APPROVED** [review]
RVI patch should be filed upstream.  Agreed on today's meeting that it will, so approved with comment. Thanks.

**Comment by leon-anavi** [comment]
Hi @zeenix, @cajun-rat,

Following the agreement during GENIVI open call about /etc/sota.toml I have rebased and updated the Yocto/OE recipe rvi-sample-config.

Best regards,
Leon

**Review by zeenix — APPROVED** [review]

**Review by zeenix — CHANGES_REQUESTED** [review]
Thanks for the quick action @leon-anavi !

Since this is not sample configuration anymore but rather the default, do we want: 

* separate recipe for adding the default config?
* this to be in a separate patch than the one that adds Aktualizer? i-e shouldn't the patches be squashed?

**Review by chbae — APPROVED** [review]

**Comment by leon-anavi** [comment]
Hi @zeenix,

You insisted to change the name and I did it but it still remains a sample configuration. As explained in the previous messages, the user has to change the device ID to match the ID from the UI. Furthermore in certain use case the user might want to configure a different server.

The Yocto/OE recipe rvi-sample-config brings not only sota.toml but also other configurations. Having all sample configurations in a separate recipe is convenient because this way they can be replaced very easily with other configurations depending on the use cause.

Yesterday, we all discussed the details related to the configuration files during the GENIVI open call and we came to an agreement which I have immediately applied in this GitHub pull request. 

After merging the GitHub pull request there are several follow-up tasks:
* Update meta-genivi-dev in GDP repo
* Remove meta-rust in GDP repo
* Upstream a fix to rvi_lib (GDP-575)
* Write Aktualizr user's guide in GENIVI wiki

I hope that our agreement from yesterday is still valid and the GitHub pull request will be merged soon because it blocks the follow-up tasks. It has to be merged before we can proceed with them.

Best regards,
Leon

**Comment by zeenix** [comment]
> You insisted to change the name and I did it but it still remains a sample configuration. 

I insisted that this be the default configuration and rename of the file was the means to achieve that, I was informed. According to my understanding, that was our agreement from yesterday's call. @cajun-rat Please do correct me if I'm wrong.

> As explained in the previous messages, the user has to change the device ID to match the ID from the UI.

I understand and that's why we'll be providing instructions for setting up a demo but as we agreed in the call yesterday, we want the SOTA on GDP to be as much working out of the box as possible and keep the number of steps still needed from the user, small.

> The Yocto/OE recipe rvi-sample-config brings not only sota.toml but also other configurations. Having all sample configurations in a separate recipe is convenient because this way they can be replaced very easily with other configurations depending on the use cause.

Good argument and I'll agree to that. However, I want this to be default configuration of GDP rather, for reasons mentioned above, hence filenames and commit log will have to reflect that. Besides you already renamed the toml file rename so it's already the default config. A deployment could then just replace the default configuration recipe (or bbappend it) with their own easily, as you intended.

> After merging the GitHub pull request there are several follow-up tasks:

I can take care of the first two for sure. :)

**Comment by cajun-rat** [comment]
@zeenix
We have to stop these cycles of repeated requests for 'just one more' additional change. Making changes to satisfy your desires for how the project and its git history looks is OK, but every time we go round this cycle it costs us real money.
What we have here works, and has gone through enough iterations over the last month so that it isn't a 'rough draft'. We'd dearly like to get on with responding to feedback from GENIVI users and making it better, but that requires getting the code in first.

**Comment by zeenix** [comment]
@cajun-rat I'm terribly sorry that you are not used to open source development, where quality is considered important. When I took over the project, I was told that GDP intends to be such a project and all my efforts (including doing proper reviews) has been part of that. You need to get used to this and take into account when estimating efforts. If you want to save time, I suggest you focus on fixing the patches, instead of trying to push an inferior development model on GDP.

Now yesterday, we had a nice open discussion about this and we agreed on going forward with my preference, not because they are 'mine' and I'm the maintainer but because that's the right way forward. You had the chance to argue how that is not the case but since you didn't argue and agreed, it is very unfair to now complain about me expecting that agreement to be followed.

BTW, how does merging this now and fixing later, rather than fixing first and then merging, drastically change the amount of efforts and time involved? The wish to merge one's efforts ASAP gives one motivation to fix the code. Leaving things for later is a major reason for all the horrible code out there in the wild.

I'd really want SOTA to shine in GDP12 and report something more substantial about it than simply that it was re-written for non-technical reasons. I hope you can agree on that and help me achieve that.

**Comment by zeenix** [comment]
Anyway, back to actual work: I'm very much meaning to merge this and I'm 100% certain that I'll do that immediately after the last changes that **we all agreed on** in yesterday's call, have been reflected in the filenames and commit log (not a matter of style, but rather correctly summarizing changes): GDP provides default configuration rather than sample.

OR if we want to merge this ASAP, we can for now remove the commit adding sample configuration and put that in a separate PR afterwards.

**Comment by gunnarx** [comment]
@zeenix I know I'm backtracking 2-3 days here but:
> this to be in a separate patch than the one that adds Aktualizer? i-e shouldn't the patches be squashed?

In the interest of time, I would propose we now accept that even if it's in a separate commit, if the commit at least has clear content and explanation.   While grouping related changes together makes it easier in those cases when you *do* have to go back and investigate the content, I believe here the ability to get the code in and do additional incremental improvements in the future will matter more for the project.

**Comment by zeenix** [comment]
@gunnarx 

1. I already agreed with @leon-anavi that squashing is not a good idea: https://github.com/GENIVI/meta-genivi-dev/pull/72#issuecomment-287009020
2. I can always help with squashing and stuff like that and that would be a good way to save time, not giving up on some important rules.

**Comment by leon-anavi** [comment]
As far as I remember during the GENIVI open call last week we all agreed to:
* Rename sota.toml.samle to sota.toml (done)
* Add a comment in sota.toml about setting the device ID (done)
* Upstream a fix to rvi_lib (reported in JIRA GDP-575 and blocked by this GitHub pull request)

I have done the required actions on my side on Wednesday last week, immediately after the call. It has been almost a week since then. Although today I read this thread again twice, I am still not sure what is blocking the merge of this GitHub pull request.

I have just updated the Git commit message for adding rvi-sample-config to further explain that the user has to set the device ID in sota.toml (as it has been already explained here in https://github.com/GENIVI/meta-genivi-dev/pull/72#issuecomment-286495667 and https://github.com/GENIVI/meta-genivi-dev/pull/72#issuecomment-286987752).

**Comment by zeenix** [comment]
@leon-anavi Thanks for your work! Although I believe I explained more than once what is left, I apologise if I didn't manage to get the message across. Let me try once more. So the steps that you mentioned (and have already acted on) were to turn the sample configuration to default configuration (which is what we agreed on in the weekly call: "Conclusion: deploy the sota sample config as default config." https://at.projects.genivi.org/wiki/display/GDP/GDP+Delivery+Report+Week+10+2017#GDPDeliveryReportWeek102017-Minutes

If that is already the case, the filenames (recipes) and commit log need to be modified since they still claim this to be sample configuration. I hope it's clear now.

**Comment by leon-anavi** [comment]
I have updated the GitHub pull request once again. I have replaced the word "sample" with "default" in the filenames, the description of the recipe rvi-default-config and in the Git commit messages. If you don't like the Git commit messages please feel free to git rebase & amend them.

**Comment by zeenix** [comment]
@leon-anavi Thanks so much. I was actually thinking of doing that for you (at least the commit log) but I was awaiting for either you or @cajun-rat to confirm if your changes indeed did turn the sample configuration to default. It was not clear since the recipes and commit log had not been updated before now. Hence the confusion and lack of help from my side.

I can surely always help you with *fixing* the commit logs at least but I would appreciate if you request that politely and not imply that my criticism is about personal preferences.

**Comment by zeenix** [comment]
@leon-anavi Do you want me to proceed with my these steps as I promised?

- Update meta-genivi-dev in GDP repo
- Remove meta-rust in GDP repo

**Comment by leon-anavi** [comment]
Thank you for merging Aktualizr. I will take care of the follow up tasks.

**Pull request merged**
